File: pkg/@eser/workflows/engine.ts
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

/**
 * Workflow execution engine.
 *
 * Pure functions that run workflow pipelines. The engine itself performs
 * no I/O — mutations are returned in results and the caller decides
 * whether to apply them (via the `onMutations` callback).
 *
 * Uses `@eser/functions/task` for lazy composition, timeout handling,
 * and result-based error recovery.
 *
 * ```
 * runWorkflow(workflow, registry, options)
 *   │
 *   ├─ For each step:
 *   │   ├─ resolveStep() → { name, options, continueOnError, timeout }
 *   │   ├─ Lookup tool in registry
 *   │   ├─ Merge step options + run options
 *   │   ├─ Run tool via task.withTimeout()
 *   │   ├─ On error: if continueOnError → failed step; else propagate error
 *   │   ├─ Call onMutations() if mutations produced
 *   │   └─ Call onStepEnd()
 *   │
 *   └─ Return aggregate WorkflowResult
 *
 * resolveIncludes(workflow, allWorkflows)
 *   ├─ For each included id, find in allWorkflows
 *   ├─ Prepend their steps to current workflow's steps
 *   ├─ Track visited ids → detect circular includes → throw
 *   └─ Return flattened WorkflowDefinition
 * ```
 *
 * @module
 */

import * as task from "@eser/functions/task";
import * as results from "@eser/primitives/results";
import type { Registry } from "./registry.ts";
import type {
  ResolvedStep,
  RunOptions,
  StepResult,
  WorkflowDefinition,
  WorkflowError,
  WorkflowResult,
  WorkflowsConfig,
  WorkflowStepConfig,
  WorkflowToolResult,
} from "./types.ts";

/**
 * Create a WorkflowError value.
 */
const workflowError = (message: string): WorkflowError => ({
  _tag: "WorkflowError",
  message,
});

/**
 * Parse a step config into a resolved step with name, options, and engine directives.
 *
 * Engine directives (`continueOnError`, `timeout`) are extracted from the options
 * and placed on the ResolvedStep. They are NOT passed to the tool.
 *
 * - `"fix-eof"` → `{ name: "fix-eof", options: {}, continueOnError: false }`
 * - `{ "check-json": { exclude: ["x"], timeout: 120 } }` → timeout extracted, options clean
 */
export const resolveStep = (step: WorkflowStepConfig): ResolvedStep => {
  if (typeof step === "string") {
    return { name: step, options: {}, continueOnError: false };
  }

  const entries = Object.entries(step);
  if (entries.length !== 1) {
    throw new Error(
      `Invalid step config: expected exactly one key, got ${entries.length}`,
    );
  }

  const [name, rawOptions] = entries[0]!;
  const options: Record<string, unknown> = {};
  let continueOnError = false;
  let timeout: number | undefined;

  for (const [key, value] of Object.entries(rawOptions)) {
    if (key === "continueOnError") {
      continueOnError = value === true;
    } else if (key === "timeout") {
      // YAML timeout is in seconds, engine uses milliseconds
      timeout = (value as number) * 1000;
    } else {
      options[key] = value;
    }
  }

  return { name, options, continueOnError, timeout };
};

/**
 * Create a Task that runs a single tool step with timeout.
 * Returns a Task<WorkflowToolResult, WorkflowError, void>.
 */
const createStepTask = (
  toolRun: () => Promise<WorkflowToolResult>,
  timeoutMs: number,
  stepName: string,
): task.Task<WorkflowToolResult, WorkflowError> =>
  task.withTimeout(
    task.fromPromise(
      toolRun,
      (error) =>
        workflowError(
          error instanceof Error ? error.message : String(error),
        ),
    ),
    timeoutMs,
    workflowError(
      `Step '${stepName}' timed out after ${(timeoutMs / 1000).toFixed(0)}s`,
    ),
  );

/**
 * Run a single workflow definition against a registry of tools.
 *
 * Returns a `Task<WorkflowResult, WorkflowError>` — the computation is
 * lazy and does not execute until `task.runTask()` is called.
 *
 * @param workflow - Workflow to run
 * @param registry - Tool registry
 * @param options - Run options
 * @returns Task wrapping the workflow execution
 */
export const runWorkflow = (
  workflow: WorkflowDefinition,
  registry: Registry,
  options: RunOptions = {},
): task.Task<WorkflowResult, WorkflowError> =>
  task.task<WorkflowResult, WorkflowError>(async () => {
    const startTime = performance.now();
    const stepResults: StepResult[] = [];
    const defaultTimeout = options.defaultTimeout ?? 60_000;

    for (const stepConfig of workflow.steps) {
      const resolved = resolveStep(stepConfig);

      // Filter by --only flag
      if (options.only !== undefined && resolved.name !== options.only) {
        continue;
      }

      // Resolve tool from registry
      const tool = registry.get(resolved.name);
      if (tool === undefined) {
        return results.fail(
          workflowError(
            `Unknown tool '${resolved.name}' in workflow '${workflow.id}'. ` +
              `Registered tools: ${registry.names().join(", ") || "(none)"}`,
          ),
        );
      }

      // Merge step-level options with run-level options (engine directives already stripped)
      const mergedOptions: Record<string, unknown> = {
        ...resolved.options,
        root: options.root ?? ".",
        fix: options.fix ?? false,
        _args: options.args ?? [],
      };

      // Pass changed files for incremental mode
      if (options.changedFiles !== undefined) {
        mergedOptions["_changedFiles"] = options.changedFiles;
      }

      // Notify step start
      options.onStepStart?.(resolved.name);

      const stepStart = performance.now();
      const timeoutMs = resolved.timeout ?? defaultTimeout;
      const stepTask = createStepTask(
        () => tool.run(mergedOptions),
        timeoutMs,
        resolved.name,
      );

      const stepTaskResult = await task.runTask(stepTask);

      let toolResult: WorkflowToolResult;

      if (results.isOk(stepTaskResult)) {
        toolResult = stepTaskResult.value;
      } else if (resolved.continueOnError) {
        // Create a failed step result from the error
        toolResult = {
          name: resolved.name,
          passed: false,
          issues: [{ message: stepTaskResult.error.message }],
          mutations: [],
          stats: {},
        };
      } else {
        return stepTaskResult;
      }

      const durationMs = performance.now() - stepStart;

      const stepResult: StepResult = {
        ...toolResult,
        durationMs,
      };

      stepResults.push(stepResult);

      // Apply mutations between steps so subsequent tools see the fixed state
      if (
        toolResult.mutations.length > 0 && options.onMutations !== undefined
      ) {
        await options.onMutations(toolResult.mutations);
      }

      // Notify step end
      options.onStepEnd?.(stepResult);
    }

    const totalDurationMs = performance.now() - startTime;

    return results.ok({
      workflowId: workflow.id,
      passed: stepResults.every((s) => s.passed),
      steps: stepResults,
      totalDurationMs,
    });
  });

/**
 * Resolve includes by flattening included workflow steps before the current workflow's steps.
 * Detects circular includes.
 *
 * @param workflow - Workflow with potential includes
 * @param allWorkflows - All available workflow definitions
 * @param visited - Set of visited workflow ids (for cycle detection)
 * @returns A new WorkflowDefinition with includes resolved (steps flattened)
 */
export const resolveIncludes = (
  workflow: WorkflowDefinition,
  allWorkflows: readonly WorkflowDefinition[],
  visited: ReadonlySet<string> = new Set(),
): WorkflowDefinition => {
  if (
    workflow.includes === undefined || workflow.includes.length === 0
  ) {
    return workflow;
  }

  const currentVisited = new Set(visited);
  currentVisited.add(workflow.id);

  const prependedSteps: WorkflowStepConfig[] = [];

  for (const includeId of workflow.includes) {
    if (currentVisited.has(includeId)) {
      throw new Error(
        `Circular include detected: workflow '${workflow.id}' includes '${includeId}' ` +
          `which is already in the include chain: ${
            [...currentVisited].join(" → ")
          }`,
      );
    }

    const included = allWorkflows.find((w) => w.id === includeId);
    if (included === undefined) {
      throw new Error(
        `Workflow '${workflow.id}' includes '${includeId}' but no workflow with that id exists. ` +
          `Available: ${allWorkflows.map((w) => w.id).join(", ") || "(none)"}`,
      );
    }

    // Recursively resolve includes of the included workflow
    const resolved = resolveIncludes(included, allWorkflows, currentVisited);
    prependedSteps.push(...resolved.steps);
  }

  return {
    ...workflow,
    steps: [...prependedSteps, ...workflow.steps],
    includes: undefined, // Resolved — no longer needed
  };
};

/**
 * Run a workflow by id from a full config, resolving includes.
 *
 * @param workflowId - Workflow id to find and run
 * @param config - Full workflow configuration
 * @param registry - Tool registry
 * @param options - Run options
 * @returns Task wrapping the workflow execution
 */
export const runWorkflowWithConfig = (
  workflowId: string,
  config: WorkflowsConfig,
  registry: Registry,
  options: RunOptions = {},
): task.Task<WorkflowResult, WorkflowError> =>
  task.task<WorkflowResult, WorkflowError>(async () => {
    const workflow = config.workflows.find((w) => w.id === workflowId);
    if (workflow === undefined) {
      return results.fail(
        workflowError(
          `Workflow '${workflowId}' not found. ` +
            `Available: ${
              config.workflows.map((w) => w.id).join(", ") || "(none)"
            }`,
        ),
      );
    }

    const resolved = resolveIncludes(workflow, config.workflows);
    return await task.runTask(runWorkflow(resolved, registry, options));
  });

/**
 * Run all workflows matching a given event, resolving includes.
 *
 * @param event - Event name to match (e.g., "precommit")
 * @param workflows - Available workflow definitions
 * @param registry - Tool registry
 * @param options - Run options
 * @returns Task wrapping an array of workflow results
 */
export const runByEvent = (
  event: string,
  workflows: readonly WorkflowDefinition[],
  registry: Registry,
  options: RunOptions = {},
): task.Task<readonly WorkflowResult[], WorkflowError> =>
  task.task<readonly WorkflowResult[], WorkflowError>(async () => {
    const matching = workflows.filter((w) => w.on.includes(event));

    if (matching.length === 0) {
      return results.fail(
        workflowError(
          `No workflows found for event '${event}'. ` +
            `Available: ${
              workflows.map((w) => `${w.id} (${w.on.join(", ")})`).join(
                "; ",
              ) || "(none)"
            }`,
        ),
      );
    }

    const workflowResults: WorkflowResult[] = [];
    for (const workflow of matching) {
      const resolved = resolveIncludes(workflow, workflows);
      const result = await task.runTask(
        runWorkflow(resolved, registry, options),
      );
      if (results.isFail(result)) return result;
      workflowResults.push(result.value);
    }

    return results.ok(workflowResults);
  });

File: pkg/ajan/workflowfx/engine.go
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

package workflowfx

import (
	"context"
	"fmt"
	"strings"
	"time"
)

// ResolveIncludes flattens a workflow's `Includes` list by prepending the steps
// of each referenced workflow. Circular references are detected and cause an error.
func ResolveIncludes(workflow *WorkflowDefinition, all []*WorkflowDefinition) (*WorkflowDefinition, error) {
	visited := make(map[string]bool)

	return resolveIncludesInner(workflow, all, visited)
}

func resolveIncludesInner(
	workflow *WorkflowDefinition,
	all []*WorkflowDefinition,
	visited map[string]bool,
) (*WorkflowDefinition, error) {
	if visited[workflow.ID] {
		return nil, fmt.Errorf("%w: '%s'", ErrCircularIncludes, workflow.ID)
	}

	visited[workflow.ID] = true
	defer func() { visited[workflow.ID] = false }()

	if len(workflow.Includes) == 0 {
		return workflow, nil
	}

	var prepended []StepConfig

	for _, id := range workflow.Includes {
		found := findWorkflow(id, all)
		if found == nil {
			return nil, &WorkflowNotFoundError{ID: id}
		}

		resolved, err := resolveIncludesInner(found, all, visited)
		if err != nil {
			return nil, err
		}

		prepended = append(prepended, resolved.Steps...)
	}

	flat := &WorkflowDefinition{
		ID:    workflow.ID,
		On:    workflow.On,
		Steps: append(prepended, workflow.Steps...),
	}

	return flat, nil
}

func findWorkflow(id string, all []*WorkflowDefinition) *WorkflowDefinition {
	for _, w := range all {
		if w.ID == id {
			return w
		}
	}

	return nil
}

// RunWorkflow executes all steps of a workflow against the provided registry.
func RunWorkflow(
	ctx context.Context,
	workflow *WorkflowDefinition,
	registry *Registry,
	opts *RunOptions,
) (*WorkflowResult, error) {
	start := time.Now()
	steps := make([]StepResult, 0, len(workflow.Steps))
	passed := true
	defaultTimeout := opts.defaultTimeout()

	// Mutable variable scope: initialized from workflow definition, updated by variable-set.
	variables := cloneStepOptions(workflow.Variables)

	// queue allows NextSteps injection and Loop expansion at runtime.
	queue := make([]StepConfig, len(workflow.Steps))
	copy(queue, workflow.Steps)

	var prev *WorkflowToolResult

	for len(queue) > 0 {
		step := queue[0]
		queue = queue[1:]

		if opts != nil && opts.Only != "" && step.Name != opts.Only {
			continue
		}

		// Build the Expr environment for this step.
		// steps contains results of all previously completed steps (0-based).
		// variables is the live mutable scope (updated by variable-set).
		env := buildExprEnv(workflow, opts, prev, steps, variables)

		// Evaluate If predicate; skip step when false.
		if step.If != "" {
			ok, err := evalBoolExpr(step.If, env)
			if err != nil {
				return nil, newWorkflowError(
					fmt.Sprintf("step '%s' if-predicate: %v", step.Name, err),
					err,
				)
			}

			if !ok {
				continue
			}
		}

		// Expand Loop into individual iterations prepended to the queue.
		if step.Loop != nil {
			items, err := evalSliceExpr(step.Loop.Over, env)
			if err != nil {
				return nil, newWorkflowError(
					fmt.Sprintf("step '%s' loop.over: %v", step.Name, err),
					err,
				)
			}

			expanded := make([]StepConfig, 0, len(items))
			for i, item := range items {
				iter := step
				iter.Loop = nil
				iter.Options = cloneStepOptions(step.Options)
				iter.Options[step.Loop.As] = item

				if step.Loop.AsIndex != "" {
					iter.Options[step.Loop.AsIndex] = i
				}

				expanded = append(expanded, iter)
			}

			queue = append(expanded, queue...)

			continue
		}

		// Bypass: skip tool execution; carry previous step's stats as output.
		if step.Bypass {
			var prevStats map[string]any
			if prev != nil && prev.Stats != nil {
				prevStats = cloneStepOptions(prev.Stats)
			} else {
				prevStats = map[string]any{}
			}

			bypassResult := &WorkflowToolResult{
				Name:      step.Name,
				Passed:    true,
				Issues:    []WorkflowIssue{},
				Mutations: nil,
				Stats:     prevStats,
			}
			sr := StepResult{WorkflowToolResult: *bypassResult, DurationMs: 0}
			steps = append(steps, sr)

			if opts != nil && opts.OnStepEnd != nil {
				opts.OnStepEnd(step.Name, sr)
			}

			prev = bypassResult

			continue
		}

		tool, ok := registry.Get(step.Name)
		if !ok {
			return nil, newWorkflowError(
				fmt.Sprintf(
					"unknown tool '%s' in workflow '%s'. registered: %s",
					step.Name, workflow.ID, strings.Join(registry.Names(), ", "),
				),
				ErrToolNotFound,
			)
		}

		rawOpts := buildStepOptions(&step, opts)

		mergedOpts, interpErr := InterpolateOptions(rawOpts, env)
		if interpErr != nil {
			return nil, newWorkflowError(
				fmt.Sprintf("step '%s' interpolation: %v", step.Name, interpErr),
				interpErr,
			)
		}

		// Input schema validation: runs on the post-interpolation merged options.
		if len(step.InputSchema) > 0 {
			if valErr := ValidateStepInput(step.InputSchema, mergedOpts); valErr != nil {
				schemaFailResult := failResult(step.Name, fmt.Sprintf("schema validation: %s", valErr.Error()))
				sr := StepResult{WorkflowToolResult: *schemaFailResult, DurationMs: 0}

				if !step.ContinueOnError {
					return nil, newWorkflowError(
						fmt.Sprintf("step '%s' input schema validation failed: %s", step.Name, valErr),
						valErr,
					)
				}

				passed = false
				steps = append(steps, sr)

				if opts != nil && opts.OnStepEnd != nil {
					opts.OnStepEnd(step.Name, sr)
				}

				prev = schemaFailResult

				continue
			}
		}

		if opts != nil && opts.OnStepStart != nil {
			opts.OnStepStart(step.Name)
		}

		timeout := step.Timeout
		if timeout == 0 {
			timeout = defaultTimeout
		}

		stepCtx, cancel := context.WithTimeout(ctx, timeout)
		stepStart := time.Now()

		toolResult, err := tool.Run(stepCtx, mergedOpts)
		cancel()

		durationMs := float64(time.Since(stepStart).Milliseconds())

		if err != nil {
			if !step.ContinueOnError {
				return nil, newWorkflowError(
					fmt.Sprintf("step '%s' failed: %v", step.Name, err),
					err,
				)
			}

			toolResult = &WorkflowToolResult{
				Name:      step.Name,
				Passed:    false,
				Issues:    []WorkflowIssue{{Message: err.Error()}},
				Mutations: nil,
				Stats:     map[string]any{},
			}
		}

		if toolResult == nil {
			toolResult = &WorkflowToolResult{
				Name:      step.Name,
				Passed:    true,
				Issues:    []WorkflowIssue{},
				Mutations: nil,
				Stats:     map[string]any{},
			}
		}

		// variable-set: promote Stats["name"]/Stats["value"] into the live variable scope.
		if step.Name == "variable-set" && toolResult.Passed {
			if name, ok := toolResult.Stats["name"].(string); ok && name != "" {
				variables[name] = toolResult.Stats["value"]
			}
		}

		// Prepend any dynamically generated steps before the remaining queue.
		if len(toolResult.NextSteps) > 0 {
			queue = append(toolResult.NextSteps, queue...)
		}

		if !toolResult.Passed {
			passed = false
		}

		stepResult := StepResult{
			WorkflowToolResult: *toolResult,
			DurationMs:         durationMs,
		}

		steps = append(steps, stepResult)

		if opts != nil && opts.OnStepEnd != nil {
			opts.OnStepEnd(step.Name, stepResult)
		}

		if len(toolResult.Mutations) > 0 && opts != nil && opts.OnMutations != nil {
			if err := opts.OnMutations(toolResult.Mutations); err != nil {
				return nil, newWorkflowError("mutation handler failed", err)
			}
		}

		prev = toolResult
	}

	return &WorkflowResult{
		ID:              workflow.ID,
		Passed:          passed,
		Steps:           steps,
		TotalDurationMs: float64(time.Since(start).Milliseconds()),
	}, nil
}

// buildStepOptions merges step-level options with run-level options.
func buildStepOptions(step *StepConfig, opts *RunOptions) map[string]any {
	merged := make(map[string]any, len(step.Options)+4)

	for k, v := range step.Options {
		merged[k] = v
	}

	root := "."
	fix := false
	args := []string{}

	if opts != nil {
		if opts.Root != "" {
			root = opts.Root
		}

		fix = opts.Fix
		args = opts.Args

		if opts.ChangedFiles != nil {
			merged["_changedFiles"] = opts.ChangedFiles
		}
	}

	merged["root"] = root
	merged["fix"] = fix
	merged["_args"] = args

	return merged
}

// RunByEvent runs all workflows whose On list includes the given event.
func RunByEvent(
	ctx context.Context,
	event string,
	workflows []*WorkflowDefinition,
	registry *Registry,
	opts *RunOptions,
) ([]*WorkflowResult, error) {
	var matching []*WorkflowDefinition

	for _, w := range workflows {
		for _, e := range w.On {
			if e == event {
				matching = append(matching, w)

				break
			}
		}
	}

	if len(matching) == 0 {
		available := make([]string, 0, len(workflows))

		for _, w := range workflows {
			available = append(available, fmt.Sprintf("%s (%s)", w.ID, strings.Join(w.On, ", ")))
		}

		return nil, fmt.Errorf(
			"%w '%s'. available: %s",
			ErrNoMatchingEvent,
			event,
			strings.Join(available, "; "),
		)
	}

	results := make([]*WorkflowResult, 0, len(matching))

	for _, w := range matching {
		resolved, err := ResolveIncludes(w, workflows)
		if err != nil {
			return nil, err
		}

		result, err := RunWorkflow(ctx, resolved, registry, opts)
		if err != nil {
			return nil, err
		}

		results = append(results, result)
	}

	return results, nil
}

// RunWorkflowWithConfig resolves and runs a workflow by ID from a full config.
func RunWorkflowWithConfig(
	ctx context.Context,
	workflowID string,
	cfg *WorkflowsConfig,
	registry *Registry,
	opts *RunOptions,
) (*WorkflowResult, error) {
	w := findWorkflow(workflowID, cfg.Workflows)
	if w == nil {
		available := make([]string, 0, len(cfg.Workflows))

		for _, wf := range cfg.Workflows {
			available = append(available, wf.ID)
		}

		return nil, fmt.Errorf(
			"%w '%s'. available: %s",
			ErrWorkflowNotFound,
			workflowID,
			strings.Join(available, ", "),
		)
	}

	resolved, err := ResolveIncludes(w, cfg.Workflows)
	if err != nil {
		return nil, err
	}

	return RunWorkflow(ctx, resolved, registry, opts)
}
